recirq/toric_code/optimizers.py

```diff
@@ -0,0 +1,460 @@
+    moment_with_czs_instead = cirq.Moment(ops_besides_cnot + czs)
+    return hadamards, moment_with_czs_instead, hadamards
+
+
+def _combine_1q_cliffords(*gates: cirq.Gate) -> cirq.Gate:
+    """Multiply 1q Clifford unitaries and simplify to a human-friendly representation.
+
+    Args:
+        gates: 1q Clifford gates in sequential order of application (index 0 first)
+
+    Returns:
+        Single 1q Clifford gate equivalent to gates. We resolve to these gates if possible:
+        I, H, X, Y, Z, X**0.5, Y**0.5, S, X**-0.5, Y**-0.5, S**-1
+        Otherwise, we return a cirq.SingleQubitCliffordGate.
+    """
+    unitary = np.asarray(cirq.linalg.dot(*(cirq.unitary(gate) for gate in gates[::-1])))
+    clifford = cirq.SingleQubitCliffordGate.from_unitary(unitary)
+
+    clifford_to_friendly_rep: Dict[cirq.SingleQubitCliffordGate, cirq.Gate] = {
+        cirq.SingleQubitCliffordGate.I: cirq.I,
+        cirq.SingleQubitCliffordGate.H: cirq.H,
+        cirq.SingleQubitCliffordGate.X: cirq.X,
+        cirq.SingleQubitCliffordGate.Y: cirq.Y,
+        cirq.SingleQubitCliffordGate.Z: cirq.Z,
+        cirq.SingleQubitCliffordGate.X_sqrt: cirq.X ** 0.5,
+        cirq.SingleQubitCliffordGate.Y_sqrt: cirq.Y ** 0.5,
+        cirq.SingleQubitCliffordGate.Z_sqrt: cirq.S,
+        cirq.SingleQubitCliffordGate.X_nsqrt: cirq.X ** -0.5,
+        cirq.SingleQubitCliffordGate.Y_nsqrt: cirq.Y ** -0.5,
+        cirq.SingleQubitCliffordGate.Z_nsqrt: cirq.S ** -1,
+    }
+    return clifford_to_friendly_rep.get(clifford, clifford)
+
+
+def convert_cnot_moments_to_cz_and_simplify_hadamards(
+    circuit: cirq.Circuit,
+) -> cirq.Circuit:
+    """Create a new circuit with CNOT replaced with CZ and simplified H.
+
+    We preserve the moment structure of the CNOTs and insert Hadamards as needed.
+    """
+    moment_on_deck = cirq.Moment()
+    new_circuit = cirq.Circuit()
+
+    for moment in circuit:
+        if _has_2q_gates(moment):
+            hadamard0, cz, hadamard1 = _break_up_cnots(moment)
+            new_circuit += _merge_hadamards(moment_on_deck, hadamard0)
+            new_circuit += cz
+            moment_on_deck = hadamard1
+        elif _is_exclusively_1q_gates(moment):
+            merge_moments = _merge_hadamards(moment_on_deck, moment)
+            if len(merge_moments) > 1:
+                new_circuit += merge_moments[0]
+            moment_on_deck = merge_moments[-1] or cirq.Moment()
+        else:
+            if moment_on_deck:
+                new_circuit += moment_on_deck
+                moment_on_deck = cirq.Moment()
+            new_circuit += moment
+
+    if moment_on_deck:
+        new_circuit += moment_on_deck
+
+    return new_circuit
+
+
+def defer_single_qubit_gates(circuit: cirq.Circuit) -> cirq.Circuit:
+    """Create a new circuit where some early single-qubit gates have been pushed later.
+
+    This is useful if the qubits begin the circuit in |0⟩; we leave them in the initial state
```

Comment:
this sentence is grammatically confusing

Reply:
Changed.